src/platformio.ts

```diff
@@ -100,6 +120,37 @@ export class PlatformIO {
         this._pioTerminal.onDidCloseTerminal(closedTerminal);
     }
 
+    private configureAutoUpdateIncludes(): void {
+        const autoUpdateIncludes = vscode.workspace.getConfiguration("platformio").get<boolean>("autoUpdateIncludes");
+        if (autoUpdateIncludes && !this._autoUpdateIncludesEnabled) {
+            const fileSystemWatcher: vscode.FileSystemWatcher = vscode.workspace.createFileSystemWatcher("**");
+            this._autoUpdateIncludesDisposables.push(fileSystemWatcher);
+            fileSystemWatcher.onDidCreate((e: vscode.Uri) => this.fileSystemDidChange(e), this, this._autoUpdateIncludesDisposables);
+            fileSystemWatcher.onDidDelete((e: vscode.Uri) => this.fileSystemDidChange(e), this, this._autoUpdateIncludesDisposables);
+            const iniFileWatcher: vscode.FileSystemWatcher = vscode.workspace.createFileSystemWatcher("**/platformio.ini");
+            this._autoUpdateIncludesDisposables.push(iniFileWatcher);
+            iniFileWatcher.onDidChange((e: vscode.Uri) => this.addIncludePath(true), this, this._autoUpdateIncludesDisposables);
+            iniFileWatcher.onDidCreate((e: vscode.Uri) => this.addIncludePath(true), this, this._autoUpdateIncludesDisposables);
+        }

```

Comment:
Yes, it will also be changed if the user just add or delete a file. There doesn't seem to be a way to make the FileSystemWatcher only watch folders and not files, so if the user add a library with many files and folders, the event will trigger many times. 

When the FileSystemWatcher triggers the event, it only supplies the path without any indication weather it is a file or folder and I'm afraid that making file system calls to find out for every event would make it slow.

It is a compromise between speed, updating too often and not updating often enough.

I have added the _excludePaths array of RegExps for paths that should not cause an update, but what if somebody have a library with a folder named "something.cpp", should we worry about that?

I'm not really happy with solutions like this, that are only almost correct - do you have a better idea?